cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/RemotePluginFinder.java

```diff
@@ -206,10 +211,42 @@ public Location getArtifactLocation(ArtifactId artifactId) throws IOException, A
 
     String path = response.getResponseBodyAsString();
     Location location = Locations.getLocationFromAbsolutePath(locationFactory, path);
+
+    // If the artifact doesnt exist locally then fetch it from app fabric and save it locally
     if (!location.exists()) {
-      throw new IOException(String.format("Artifact Location does not exist %s for artifact %s version %s",
-                                          path, artifactId.getArtifact(), artifactId.getVersion()));
+      LOG.debug(String.format("Artifact '%s' is not present locally at %s, attempting to fetch it from app-fabric.",
+                              artifactId.getArtifact(), path));
+      getAndStoreArtifact(artifactId, location);
     }
     return location;
   }
+
+  private void getAndStoreArtifact(ArtifactId artifactId, Location location) throws IOException {
+    HttpResponse httpResponse;
+    String namespaceId = artifactId.getNamespace();
+    ArtifactScope scope = ArtifactScope.USER;
+    // Cant use 'system' as the namespace in the request because that generates an error, the namespace doesnt matter
+    // as long as it exists. Using default because it will always be there
+    if (ArtifactScope.SYSTEM.toString().equalsIgnoreCase(namespaceId)) {
+      namespaceId = "default";
+      scope = ArtifactScope.SYSTEM;
+    }
+    String url = String.format("namespaces/%s/artifacts/%s/versions/%s/download?scope=%s",
+                               namespaceId,
+                               artifactId.getArtifact(),
+                               artifactId.getVersion(),
+                               scope);
+
+    LOG.debug("Fetching artifact from " + url);
+    HttpRequest.Builder requestBuilder = remoteClientInternal.requestBuilder(HttpMethod.GET, url);
+    httpResponse = remoteClientInternal.execute(requestBuilder.build());
+    if (httpResponse.getResponseCode() != HttpURLConnection.HTTP_OK) {
+      throw new IOException(String.format("Failed to fetch artifact %s version %s from app-fabric",
+                                          artifactId.getArtifact(), artifactId.getVersion()));
+    }
+    OutputStream outputStream = location.getOutputStream();
+    ByteStreams.copy(new ByteArrayInputStream(httpResponse.getResponseBody()), outputStream);
+    outputStream.close();
```

Comment:
Don't call close like this. Always use the automatic resource release.

```
try (OutputStream os = location.getOutputStream()) {
  // Use the os here.
}
```

Reply:
done